noir-projects/noir-contracts/contracts/escrowable_token_contract/src/main.nr

```diff
@@ -0,0 +1,571 @@
+     */
+    #[aztec(private)]
+    fn unshield(
+        from: AztecAddress,
+        to: AztecAddress,
+        amount: Field,
+        nonce: Field,
+        change_incoming_viewer_and_nullifier: AztecAddressOption,
+        outgoing_viewer: AztecAddressOption
+    ) {
+        let from_roles = storage.roles.at(from).get_current_value_in_private();
+        assert(!from_roles.is_blacklisted, "Blacklisted: Sender");
+        let to_roles = storage.roles.at(to).get_current_value_in_private();
+        assert(!to_roles.is_blacklisted, "Blacklisted: Recipient");
+
+        if (!from.eq(context.msg_sender())) {
+            assert_current_call_valid_authwit(&mut context, from);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        storage.balances.sub(
+            from,
+            U128::from_integer(amount),
+            change_incoming_viewer_and_nullifier.unwrap_or(from),
+            outgoing_viewer.unwrap_or(from)
+        );
+
+        EscrowableToken::at(context.this_address())._increase_public_balance(to, amount).enqueue(&mut context);
+    }
+
+    /**
+     * Transfers `amount` from the private balance of `from` to the private balance of `to`.
+     *
+     * Support for escrowing by explicitly specifying the `incoming_viewer_and_nullifier` and `outgoing_viewer` addresses.
+     * 
+     * @param from: The address to transfer from
+     * @param to: The address to transfer to
+     * @param amount: The amount to transfer
+     * @param nonce: The nonce of the transfer
+     * @param change_incoming_viewer_and_nullifier: The address to use for `ivpk` and `npk` for change notes - if not provided, the `from` address is used
+     * @param to_incoming_viewer_and_nullifier: The address to use for `ivpk` and `npk` for the to notes - if not provided, the `to` address is used
+     * @param outgoing_viewer: The address to use for `ovpk` when creating outgoing logs - if not provided, the `from` address is used
+     */
+    #[aztec(private)]
+    fn transfer(
+        from: AztecAddress,
+        to: AztecAddress,
+        amount: Field,
+        nonce: Field,
+        change_incoming_viewer_and_nullifier: AztecAddressOption,
+        to_incoming_viewer_and_nullifier: AztecAddressOption,
+        outgoing_viewer: AztecAddressOption
+    ) {
+        let from_roles = storage.roles.at(from).get_current_value_in_private();
+        assert(!from_roles.is_blacklisted, "Blacklisted: Sender");
+        let to_roles = storage.roles.at(to).get_current_value_in_private();
+        assert(!to_roles.is_blacklisted, "Blacklisted: Recipient");
+
+        if (!from.eq(context.msg_sender())) {
+            assert_current_call_valid_authwit(&mut context, from);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        let amount = U128::from_integer(amount);
+
+        storage.balances.sub(
```

Comment:
imagine a staking contract.
When user stakes there: tokenNote should still belong to user. But balance should be increased of staking contract
i.e. `balances.at(staking_contract) += amount`
SAID ANOTHER WAY: `balance[staking_contract] = TokenNote<owner = user. amount = 100>`

Later, when a user unstakes:
the staking contract should reduce its balance. Crucuially though, `get_notes` should be called on the `user` and not the `staking contract` since the note.owner = user.

So there should be a from_nullifioor. My hypothesis is there is never a case when `from_nullifioor` != `outgoing_viewoor` so you should `get_notes()` from that party and not `from`